### PR TITLE
WIP: Support of task execution's started_at and finished_at

### PR DESCRIPTION
**Pull request by vgvoleg**
Display task execution's duration, calculated by 2 new fields:
started_at and finished_at, if they exist.

Signed-off-by: Oleg Ovcharuk <vgvoleg@gmail.com>

### CODE REVIEW

**Comment by vgvoleg** [comment]
This pull request could not be merged until this [blueprint](https://blueprints.launchpad.net/mistral/+spec/mistral-add-started-finished-at) is merged

**Comment by GuyShaanan** [comment]
Hi @vgvoleg Thank you so much for your contribution.
Once the blueprint is approved and we have a Mistral running with that feature I'll be happy to review this PR with you.

**Comment by MosheElisha** [comment]
Thanks, @vgvoleg !

One important note from my side - this change will make CloudFlow work only with Mistral >= Stein and will not work with Mistral <= Rocky anymore.

Is there an easy option to make this in such a way that CloudFlow will still work even if the new fields do not exist in the response? (e.g., fallback to created_at and updated_at)

**Comment by GuyShaanan** [comment]
@MosheElisha He did just right, allowing a fallback if the values does not exist.
It seems we won't need a new version for Mistral >= Stein.

**Comment by vgvoleg** [comment]
Hi guys! This feature is merged to mistral's master.
But I've got a confused moment and I need your opinion.
Mistral's endpoint /executions/{id}/tasks with field filtering throws an http error, if any mentioned field is not exist, so, really, this PR would not work with Mistral<stein.
So, there are two possible road to fix this: to allow Mistral to return only existing fields while using field filtering (which, probably, would not be a good idea), or to double-check it on CloudFlow's side. Using the second way will result in two requests at the worst case, so if someone use Mistral<stein, task executions loading time will increased by small value (field validation process is one of the first). 
If the second way is acceptable for you, please give me advise how to implement this correctly. I'm not so good at promises, and my implementation could look very ugly :D

**Comment by GuyShaanan** [comment]
@rakhmerov Should Mistral throw an error is such case? Or just return the entity with the fields it did find?
@rakhmerov / @MosheElisha Do we still want to see the 'created_at' and 'updated_at' fields in CloudFlow alongside the new 'started_at' and 'finished_at'? If so, which are more important?

I would rather not perform 2 requests just to identify what the server supports.
Is any new version / breaking change to CloudFlow seems reasonable for this feature?

**Comment by MosheElisha** [comment]
Hi,

I think the new started_at and finished_at are very important and showed be shown alongside the current created_at' and 'updated_at' fields.

I think the current Mistral behavior of throwing an exception in case of unknown fields is correct.

Option #2 is OK by me but I am also not completely ruling out option #3 - New CloudFlow versions will work only for Mistral >= Stein.

**Comment by MosheElisha** [comment]
@vgvoleg - Thank you for the initial PR. What is the status of this PR? Will you continue or should we try to find someone else?

Thanks.

**Comment by vgvoleg** [comment]
@MosheElisha I'd like to finish this, but I didn't understand should we update supported Mistral's version or we need some refactor

**Comment by MosheElisha** [comment]
Cool! I'm OK that the new CloudFlow release will support only Mistral >= Stein.

**Comment by MosheElisha** [comment]
@vgvoleg  / @rakhmerov - should we also display the created_at and update_at timestamps so we could easily see overhead if exists?

**Comment by vgvoleg** [comment]
@MosheElisha I don't know if this is really needed: updated_at == finished_at. I think it is interesting to add just created_at to the runtimes view to see how much time tasks are in waiting state, but I don't know how to place it in view. If you have any ideas, let me know, I'd like to improve my frontend skills :D

**Comment by vgvoleg** [comment]
@MosheElisha also if it is not necessary - I can delete fallback sections from my MR and remove WIP status :)

**Comment by MosheElisha** [comment]
If updated_at is always equal to finished_at - why do we have both? :-)

Anyway, one idea is to have a check box "Use started_at" in the Runtime View that toggles between the "created_at" and the "started_at".

In the info panel on the right, we should add the "started_at" somewhere. Also think what it means regarding "Duration" field.

@GuyShaanan / @rakhmerov - thoughts?

**Comment by GuyShaanan** [comment]
Yes, I believe showing both fields sets is the way to go.

But indeed we should ask what is the meaning of _Duration_ now. Should it be adjusted to the diff `finished_at - started_at`? (Today it is https://github.com/nokia/CloudFlow/blob/907f132f44ec147e531c33d34fd740c1bce78da0/src/app/shared/models/execution.ts#L46)

**Comment by vgvoleg** [comment]
@GuyShaanan It depends. Difference between `finished` and `started` shows the execution time, difference between `started` and `created` shows how much time task was in waiting state. `updated` is pretty useless, it equals to `finished` since the task was complete.

**Comment by MosheElisha** [comment]
I understand that `updated` and `finished` are the same now but in theory we could have made `finished` more meaningful.
The `updated` marks the time that the task was updated in the DB by the engine.
This is not the same as when the execution finished by the executor or engine (depending on if it is in remote or local executor mode). 
After execution finishes, the result needs to be returned on the AMQP and only than it is updated in the DB.
Basically, I would expect that `finished` will usually be (at least slightly) earlier than the `updated`.

**Comment by vgvoleg** [comment]
@MosheElisha @GuyShaanan @rakhmerov 

I've sent a patch with some changes. Sorry for waiting, I did it as quickly as I could.  
Here's my vision of how to combine them and make both of these fields useful:  
image  
So, if there is a difference between started and created / finished and updated, we will see a waiting section on the task runtime before / after task runtime.  
Also, this is how I combine them in the task-info:  
image  
(there is an ugly shift, I'm not good at UI, but I'll try to fix it)

Please let me know what you think about this.  I'd like to discuss it and to implement the most suitable solution.

**Comment by MosheElisha** [comment]
@vgvoleg This looks great! Very nice idea regarding the runtime view!

I will let @GuyShaanan review the code itself but he can do it tomorrow as today is his birthday. Congratz, Guy!

**Comment by vgvoleg** [comment]
@GuyShaanan happy birthday!

File: src/app/executions/task-info/task-info.component.html [hunk 1]
```diff
@@ -49,6 +49,19 @@
...
+      <!-- started -->
+      <cf-info-item class="flex-1-1-auto" [prop]="properties['started_at']"></cf-info-item>
+
+      <!-- finished -->
+      <cf-info-item class="flex-1-1-auto" [prop]="properties['finished_at']"></cf-info-item>
+
+      <!-- duration -->
+      <cf-info-item class="flex-1-1-auto" [prop]="properties['duration']"></cf-info-item>
+
+    </div>
```
Comment (GuyShaanan) [inline src/app/executions/task-info/task-info.component.html]: A quick solution (although not perfect) is to set a base width for the items (with a width of 33% each) so they'll have to wrap but still keep the spacing.

1) Move all 6 `cf-info-item` (of started, finished, duration, created_at, updated_at, duration_total) inside one div (i.e. delete lines 63-65)
2) Create a new style in `styles.scss`:
```scss
.flex-1-1-33 {
  flex: 1 1 33%;
}
```
3) Change the `cf-info-item` class to that new class (i.e. 
```html
<cf-info-item class="flex-1-1-33" [prop]="properties['started_at']"></cf-info-item>
```
)

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts [hunk 2]
```diff
@@ -36,30 +41,42 @@ export class TasksRuntimeComponent implements OnInit {
...
+    private sort = {by: STARTED_AT_RELATIVE, dir: 'asc'};
 
     ngOnInit(): void {
         this.calculateTimes(this.execution, this.tasks);
     }
 
     private createGraphModel(tasks: TaskExec[], workflowStarted: string) {
         return tasks.map(task => {
+
+            let created_at = task.created_at;
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.ts]: You can use the new way of [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring):
```javascript
const {created_at, updated_at, started_at, finished_at} = task;
```

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts [hunk 3]
```diff
@@ -92,6 +105,18 @@ export class TasksRuntimeComponent implements OnInit {
                 preBarWidth = 99;
             }
             task.preBarWidth = preBarWidth;
+
+            const waitingBeforeBarWidth = (task.started_at_relative / workflowDurationSec) * 100;
+            task.waitingBeforeBarWidth = waitingBeforeBarWidth || 0;
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.ts]: There is no benefit in doing `expr || 0` because 0 evaluates to `false`. So whatever the value of `waitingBeforeBarWidth` is, it is the one that will be taken.
So these 2 lines could be minified to
```js
task.waitingBeforeBarWidth = (task.started_at_relative / workflowDurationSec) * 100;
```

(the reason for the `runningBarWidth || 1` on line 114 is to have a minimum bar width of 1 when the running time is 0, for visual purposes)

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts [hunk 4]
```diff
@@ -92,6 +105,18 @@ export class TasksRuntimeComponent implements OnInit {
...
+
+            // set the graph's width
+            const runningBarWidth = (task.duration_sec / workflowDurationSec) * 100;
+            task.runningBarWidth = runningBarWidth || 1;
+
+            const waitingAfterBarWidth = (task.finished_at_relative / workflowDurationSec) * 100;
+            task.waitingAfterBarWidth = waitingAfterBarWidth || 0;
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.ts]: Same as above

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts [hunk 5]
```diff
@@ -92,6 +105,18 @@ export class TasksRuntimeComponent implements OnInit {
                 preBarWidth = 99;
             }
             task.preBarWidth = preBarWidth;
+
+            const waitingBeforeBarWidth = (task.started_at_relative / workflowDurationSec) * 100;
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.ts]: I would add some description to the doc string of the function, explaining the new elements you've added.

File: src/app/executions/tasks-runtime/tasks-runtime.component.html [hunk 6]
```diff
@@ -24,13 +24,16 @@
...
-            Percent: {{task.duration_sec === 0 ? '< ' : ''}}{{task.barWidth | number}}%</ng-template>
+          <ng-template #runtimeTooltip>Started: {{task.started_at}}<br/>
+            Finished: {{task.finished_at}}<br/>
+            Percent: {{task.duration_sec === 0 ? '< ' : ''}}{{task.runningBarWidth | number}}%<br/>
+            Total: {{task.duration_sec === 0 ? '< ' : ''}}{{task.totalBarWidth | number}}%</ng-template>
 
           <div class="inner">
             <div class="prehighlight" [style.width.%]="task.preBarWidth"></div>
-            <div class="highlight" [style.width.%]="task.barWidth"></div>
+            <div class="waiting-highlight" [style.width.%]="task.waitingBeforeBarWidth"></div>
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.html]: See comment below - we can save some css code by setting 2 classes on this item `classes="waiting highlight"`. (Same for line 36)

File: src/app/executions/tasks-runtime/tasks-runtime.component.scss [hunk 7]
```diff
@@ -13,7 +13,7 @@
     background: #dde4e6;
 
     .highlight {
-      border-radius: 2px;
+      //border-radius: 1px;
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.scss]: 1) Remove unused style
2) We can have a combined rule for 'highlight' now that we separated to 2 classes:
```scss
.highlight {      
      transition: filter .3s ease-in-out;
      background-color: $entity-completed;

      &:hover {
        filter: brightness(1.1);
      }

      &.waiting {
        background-color: $entity-waiting;
      }
    }
```
so you could delete the `.waiting-highlight` rule as well.

File: src/app/executions/tasks-runtime/tasks-runtime.component.scss [hunk 8]
```diff
@@ -22,6 +22,16 @@
       }
     }
 
+    .waiting-highlight {
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.scss]: delete

Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.scss]: done

File: src/app/executions/tasks-runtime/tasks-runtime.component.scss (hunk 7 above)
Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.scss]: done

File: src/app/executions/tasks-runtime/tasks-runtime.component.html (hunk 6 above)
Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.html]: done

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts (hunk 5 above)
Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.ts]: done, but not sure if I did it correct

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts (hunk 4 above)
Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.ts]: done

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts (hunk 2 above)
Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.ts]: done

File: src/app/executions/tasks-runtime/tasks-runtime.component.ts (hunk 3 above)
Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.ts]: TY, now I understand

File: src/app/executions/task-info/task-info.component.html (hunk 1 above)
Comment (vgvoleg) [reply src/app/executions/task-info/task-info.component.html]: Maybe I understand something wrong, but I can't do it with placing all `cf-info-item` elements to the one div. I saved the old way and changed style. Everything looks nice.
image

**Comment by vgvoleg** [comment]
@GuyShaanan everything is ready. By the way, I changed docker base version due to some jessie's dependencies issues.

File: proxy.conf.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
 {
   "/api": {
     "target": {
-      "host": "<IP>",
+      "host": "localhost",
```
Comment (GuyShaanan) [inline proxy.conf.json]: I wouldn't push that change

File: src/app/executions/tasks-runtime/tasks-runtime.component.scss [hunk 10]
```diff
@@ -13,15 +13,27 @@
...
       &:hover {
         filter: brightness(1.1);
       }
+
+      &.waiting {
+        background-color: $entity-waiting;
+      }
     }
 
+    //.waiting-highlight {
```
Comment (GuyShaanan) [inline src/app/executions/tasks-runtime/tasks-runtime.component.scss]: Delete dead code

Comment (vgvoleg) [reply src/app/executions/tasks-runtime/tasks-runtime.component.scss]: my fault, done

File: proxy.conf.json (hunk 9 above)
Comment (vgvoleg) [reply proxy.conf.json]: my fault, done

**Comment by vgvoleg** [comment]
@GuyShaanan I've updated Mistral version from Pike to Stein in MDs

File: README.md [hunk 11]
```diff
@@ -26,10 +26,9 @@ A workflow visualization tool for [OpenStack Mistral](https://github.com/opensta
 
 ## Requirements
 
-### Mistral >= Pike
-CloudFlow requires Mistral **Pike** or greater, as we rely on
-new [runtime_context](https://docs.openstack.org/developer/mistral/developer/webapi/v2.html#tasks)
-added to Mistral Pike.
+### Mistral >= Stein
```
Comment (GuyShaanan) [inline README.md]: I actually wouldn't change that (as older versions of CloudFlow support older versions of Mistral) but add a table underneath it instructing the users which version to use, like:

| Mistral Version | CloudFlow Version to use |
| -- | -- |
| Stein (or newer) | Latest stable |
| Pike - Rocky | [0.6.4](https://github.com/nokia/CloudFlow/releases/tag/v0.6.4) |

------------
Also, please manually bump the version to `0.7.0` in [package.json](https://github.com/nokia/CloudFlow/blob/master/package.json#L3) and [about.html](https://github.com/nokia/CloudFlow/blob/master/src/app/about/about.component.html#L4) and include those in this PR as well.

(That's the last thing, I swear :))

Comment (vgvoleg) [reply README.md]: Done

**Comment by MosheElisha** [comment]
Thanks, @vgvoleg and @GuyShaanan ! 
I will update our online CloudFlow (http://yaqluator.com:8000) soon,
